### PR TITLE
Decode xlink:href attributes when reading METS

### PR DESCRIPTION
**Pull request by jrwdunham**
A previous commit introduced URL-encoding of xlink:href values when writing METS. This commit introduces the corresponding URL-decoding when reading METS. Also adds some simple tests to display and confirm expected behaviour.

Connected to archivematica/Issues#206

### CODE REVIEW

File: metsrw/__init__.py [hunk 1]
```diff
@@ -43,12 +45,40 @@
...
+    'report_string',
+    'schematron_validate',
+    'sct_report_string',
+    'set_feature_broker_to_default_state',
+    'urldecode',
+    'urlencode',
+    'validate',
+    'xsd_error_log_string',
+    'xsd_validate',
+]
```
Comment (ross-spencer) [inline metsrw/__init__.py]: Learning a bit about this recently, this seems like good practice.

File: metsrw/fsentry.py [hunk 2]
```diff
@@ -337,8 +337,13 @@ def serialize_filesec(self):
...
             # Setting manually so order is correct
-            flocat.set(
-                utils.lxmlns('xlink') + 'href', utils.urlencode(self.path))
+            try:
+                flocat.set(
+                    utils.lxmlns('xlink') + 'href', utils.urlencode(self.path))
+            except ValueError:
+                raise exceptions.MetsError(
+                    'Value "{}" (for attribute xlink:href) is not a valid'
+                    ' URL.'.format(self.path))
```
Comment (ross-spencer) [inline metsrw/fsentry.py]: This seems sensible too.

File: metsrw/metadata.py [hunk 3]
```diff
@@ -277,8 +283,13 @@ def serialize(self):
...
         if self.target:
-            el.attrib[utils.lxmlns('xlink') + 'href'] = \
-                utils.urlencode(self.target)
+            try:
+                el.attrib[utils.lxmlns('xlink') + 'href'] = \
+                    utils.urlencode(self.target)
+            except ValueError:
+                raise exceptions.MetsError(
+                    'Value "{}" (for attribute xlink:href) is not a valid'
+                    ' URL.'.format(self.target))
```
Comment (ross-spencer) [inline metsrw/metadata.py]: If I read this correctly, `MetsError` is the base exception for the module. `ParseError` is the exception that might normally be raised when reading a METS file. I think this is fine, but something like `MetsEncodeError` might read clearer for exceptions where we're encoding data. I think I prefer the verb-y-ness so I think it's just a preference thing.

File: metsrw/utils.py [hunk 4]
```diff
@@ -1,5 +1,5 @@
 # -*- coding: utf-8 -*-
-from six.moves.urllib.parse import quote_plus, urlparse, urlunparse
+from six.moves.urllib.parse import quote_plus, unquote_plus, urlparse, urlunparse
```
Comment (ross-spencer) [inline metsrw/utils.py]: I think this just creeps over the line-length limit.

File: metsrw/utils.py [hunk 5]
```diff
@@ -37,12 +37,29 @@ def lxmlns(arg):
...
+    https://tools.ietf.org/html/rfc3986#section-2.1.
+    """
+    return _urlendecode(url, lambda val: quote_plus(val, safe='/'))
+
+
+def urldecode(url):
+    """Decode percent encoding introduced per RFC3986
+    https://tools.ietf.org/html/rfc3986#section-2.1.
+    """
+    return _urlendecode(url, unquote_plus)
```
Comment (ross-spencer) [inline metsrw/utils.py]: Clever!

File: tests/test_utils.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    'http://foo[bar.com/hello[1].pdf',
+)
+
+
+def test_url_encoding():
+    for url in GOOD_PATHS_SLASH_URLS:
+        assert url == metsrw.urldecode(metsrw.urlencode(url))
+    for url in BAD_URLS:
+        with pytest.raises(ValueError):
+            metsrw.urlencode(url)
```
Comment (ross-spencer) [inline tests/test_utils.py]: There are useful tests to have. Thanks for adding them Joel.

File: metsrw/metadata.py (hunk 3 above)
Comment (jrwdunham) [reply metsrw/metadata.py]: 👍

File: metsrw/utils.py (hunk 4 above)
Comment (jrwdunham) [reply metsrw/utils.py]: 👍

**Pull request merged**